### ISSUE TITLE
Proposal: Vector handling with extension(pgvector)

### ISSUE THREAD

**Issue by iitaejeong** [apache/age#1121]
**Is your feature request related to a problem? Please describe.**
Vector usage is very skyrocketing now. so, market wants to handle with them using database. but we didnt. exactly, the vectorDB has many popularity from resource 'db-engine rank'. 

**related to a problem**

but , Apache age doesn't have the function for vector engineering. so many other person who wants to apply the vector are just using the python library itself for supporting vector handling. 

**Describe the solution**

### Extension PGvector.
```
CRETATE EXTENSION vector;

SELECT embedding <-> '[3,1,2]' AS distance FROM items;
SELECT (embedding <#> '[3,1,2]') * -1 AS inner_product FROM items;
SELECT 1 - (embedding <=> '[3,1,2]') AS cosine_similarity FROM items;
```
and also can indexing function for searching of nearest neighbor. There are many very good functions except for the aforementioned functions.

**Additional context**
Except for above problem, we expand this feature to ML practitioner who wants data efficient management. the overview of additional context is compatible with Pytorch_geometric remote backend function. They builds the class for ease to integration others database. if previous problem well solve , then we might do this future work.

![image](https://github.com/apache/age/assets/129734402/48b23f0a-7fe9-4667-b723-f6ed3a98c4dc)

in below reference , it has the function for convenience. 

[Scaling Up GNNs via Remote Backends]
https://pytorch-geometric.readthedocs.io/en/latest/advanced/remote.html

**Comment by titoausten** [comment]
This will be a great addition. I understand that Graph data has to be converted into vector via graph embeddings before being passed to ML algorithms. 
This feature will very effective for AGE and attract alot of ML/DL practitioners. I really look forward to a project of this nature to contribute to.

**Comment by stdweird** [comment]
i guess that ideally we can use eg vector search (but maybe also bm25 via pg_search) to be able to form graphs connecting similar entities (instead of exact matching ones). 
eg when building a knowlegde graph for RAG, the entities and relations are harvested using eg LLMs, leading to similar entities (instead of exactly teh sme ones) because eg the source data is written by diffferent people talking about the same things, yet using different words. lacking a formal verification/cleanup process, this leads to more then one entity in the graph.

**Comment by mingfang** [comment]
> i guess that ideally we can use eg vector search (but maybe also bm25 via pg_search) to be able to form graphs connecting similar entities (instead of exact matching ones). eg when building a knowlegde graph for RAG, the entities and relations are harvested using eg LLMs, leading to similar entities (instead of exactly teh sme ones) because eg the source data is written by diffferent people talking about the same things, yet using different words. lacking a formal verification/cleanup process, this leads to more then one entity in the graph.

Unfortunately pg_search(https://github.com/paradedb/paradedb) does not work with `agtype` fields.

**Comment by stdweird** [comment]
@mingfang and how do you use this in your graph queries? 
tbh, i also don't know if bm25 makes a lot of sense when comparing to embedding vectors. the entities are only a few words typically, so not sure the vector search would miss things bm25 would not miss (this is different in RAG where a single vector represents a whole sentence or paragraph).

**Comment by mingfang** [comment]
First I have to create a function to effective do the `agtype` conversion
```sql
CREATE OR REPLACE FUNCTION public.movies_view_bm25_search(search agtype) RETURNS agtype AS $$
	SELECT id::agtype
	FROM movies_view_bm25.search(search::text)
$$ LANGUAGE sql;
```

Then you can call the search from cypher
```cypher
with public.movies_view_bm25_search('plot:drink') as it
match (v1:MOVIE)
where id(v1)=it
return v1
```

**Comment by tan-yong-sheng** [comment]
Hi, just wanna follow up on this... I wish to try both postgres's extension: `apache age` with `pgvector`... Is the docker-compose image working for both extensions? Thanks, and hope for reply.

**Comment by alph486** [comment]
Would love to see native support and / or examples for this on the site. Neo4j has their vector implementation and it would probably increase adoption to see this more widely used/publicised.

**Comment by KnowledgeGarden** [comment]
See
https://github.com/dgg32/age_vector

**Comment by clxyder** [comment]
Is this still being considered?

**Comment by connektiv8** [comment]
I'd also like to see this feature implemented as I'm building out an AI-enabled medical practice management system, so having these features in one place and able to use vectors with age would certainly be beneficial, especially when you have an AI trawling all prior letters and documents in vector format for the most appropriate language adaptation based on the writing styles of doctors, etc.

**Comment by la3lma** [comment]
This would be awesome. Vector & graph powered RAG is a very useful, pattern and getting this through Postgres would be a wonderful thing!

**Comment by qdrddr** [comment]
PGVector + PgVectorScale
Seams to be a good companion, especially combined with AGE.

**Comment by saminahbab** [comment]
sooooo.. whats the deal? 
should we expect sometime soon?

**Comment by qdrddr** [comment]
Here is how to build Docker container with both AGE + PG_Vector extentions
From https://medium.com/@mrsirsh/cloudnativepg-age-and-pg-vector-on-a-docker-image-step-1-ef0156c78f49
```Dockerfile
# First step is to build the the extension
FROM debian:bullseye-slim as builder

# use the advice from here https://cloudnative-pg.io/blog/creating-container-images/
# but also add in the deps for, and use, the AGE repo (https://age.apache.org/getstarted/quickstart/)
# + PG15 did not seem to work due to a string types issue ?? but i just bumped to 16 and it was fine
RUN set -xe ;\
    apt update && apt install wget lsb-release gnupg2 -y ;\
    sh -c 'echo "deb https://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main" > /etc/apt/sources.list.d/pgdg.list' ;\
    wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - ;\
    apt-get update ;\
    apt-get install -y postgresql-server-dev-16 build-essential git libreadline-dev zlib1g-dev flex bison; \
    cd /tmp; \
    git clone https://github.com/apache/age.git; \
    cd /tmp/age; \
    PG_CONFIG=/usr/lib/postgresql/16/bin/pg_config make; \
    PG_CONFIG=/usr/lib/postgresql/16/bin/pg_config make install

# Second step, we build the final image
FROM ghcr.io/cloudnative-pg/postgresql:16.4

# To install any package we need to be root user
USER root

# But this time we copy the .so file and .control and addme script from the build process
COPY --from=builder /usr/lib/postgresql/16/lib/age.so /usr/lib/postgresql/16/lib/
COPY --from=builder /usr/share/postgresql/16/extension/age.control /usr/share/postgresql/16/extension/
COPY  --from=builder /usr/share/postgresql/16/extension/age--1.5.0.sql /usr/share/postgresql/16/extension/

RUN set -xe; \
    apt-get update; \
    apt-get install -y --no-install-recommends \
    "postgresql-16-pgvector" ; \
    rm -fr /tmp/* ; \
    rm -rf /var/lib/apt/lists/*;

# Change the uid of postgres to 26
RUN usermod -u 26 postgres
USER 26
```

**Comment by qdrddr** [comment]
Relates to https://github.com/apache/age/pull/2088

**Comment by davidcsisk** [comment]
This would support GraphRAG for Generative AI applications (vector search combined with knowledge graph query), which helps keep PostgreSQL relevant to current technologies. There's also plenty of non-AI use cases for having both of these (knowledge graph and vector search/query) available and working well together.  My vote is "do it".

**Comment by j-hartshorn** [comment]
FWIW this would be a great addition!

**Comment by saminahbab** [comment]
+100, absolutely

**Comment by MuhammadTahaNaveed** [comment]
PRs https://github.com/apache/age/pull/2088 and https://github.com/apache/age/pull/2172 (in review) improve the interoperability of age with other extensions. For now, the regression tests for [pgvector](https://github.com/apache/age/blob/f47340d22b023d227d84aabaaa8b3061b7cf5fd7/regress/sql/pgvector.sql) can be referred to for using age with pgvector.

PS: I’ll try to come up with a blog or some documentation for this once https://github.com/apache/age/pull/2172 is merged.

**Comment by nicholasveloso** [comment]
> PRs [#2088](https://github.com/apache/age/pull/2088) and [#2172](https://github.com/apache/age/pull/2172) (in review) improve the interoperability of age with other extensions. For now, the regression tests for [pgvector](https://github.com/apache/age/blob/f47340d22b023d227d84aabaaa8b3061b7cf5fd7/regress/sql/pgvector.sql) can be referred to for using age with pgvector.
> 
> PS: I’ll try to come up with a blog or some documentation for this once [#2172](https://github.com/apache/age/pull/2172) is merged.

I hope this bloc post comes out soon. 

Also, with this will it be possible to use pgvector data types like vector, sparsevec, halfvec and bitvec data types, in Apache AGE's nodes and links properties? 

For example, if I want to make a graph of the documents in the knowledge base tables that have their documents chuncked and the vector embeddings generated, so I would like to have a graph with nodes labeled **DOCUMENT**, with properties **document_id**, **document_name**, **document_summary**, etc. and each embedded chunk would be anothet node with label **CHUNK**, with properties **chunk_id**, **chunk_content**,** embedding_vector**, **previous_chunk_id**, **next_chunk_id**. 

The last time I've tried this was quite some time ago, somewhere past the middle of last year, but at that time Apache AGE did not accept pgvector's data types, such as vector type, as the value of the **embedding_vector** property key, it returned an error that essentialy stated that the vector data type was not a valid data type and that it wasn't supported by Apache AGE.

It would be great if Apache AGE accepted pgvector data types, it would also be great if it can also generate embeddings of thr graphs (graph embeddings) to use with AI.

**Comment by imsakg** [comment]
https://docs.vectorchord.ai/vectorchord/getting-started/overview.html

This would better suit instead of pgvector

**Comment by S0LERA** [comment]
Any news? Is this at least being considered? I think it could be very useful for memory ecosystems for AI agents.

**Comment by PJerkins-MRO** [comment]
Would like too add my vote for this. Having an official docker image would be very helpful for those looking to create hybrid RAG solutions.

**Comment by jrgemignani** [comment]
Apache AGE supports the pgvector extension

**Comment by webcoderz** [comment]
do we have any docs on how to do vector search etc?

**Comment by jrgemignani** [comment]
@webcoderz AGE specific? Not that I'm aware. I would look at the regression test for it `regress/sql/pgvector.sql` for ideas and any documentation for pgvector. Furthermore, if you learn of any or create any of your own, please feel free to contribute them :)

**Comment by webcoderz** [comment]
Ahh good idea @jrgemignani yea I'll do that I was going to play around today/tomorrow on it. Will prob post what I come up with here

**Comment by webcoderz** [comment]
WIP as i mess with it more

# Apache AGE + pgvector 


- **Apache AGE** (`age`) for Cypher queries
- **pgvector** (`vector`) for vector operations + ANN indexes


## Required session setup

AGE requires loading the extension and setting `search_path`:

```sql
LOAD 'age';
SET search_path = ag_catalog, public;
```

Why `public`?
- `vector` is installed into the `public` schema in our stack
- Cypher queries that cast to `::vector` (or call distance functions) need `public` visible
- vector ops fail when `public` isn’t in the `search_path`, and start working once it is.

## Basic Cypher execution from SQL

AGE is called through `cypher(graph_name, $$ ... $$)` and you must provide an `AS (...)` return type list:

```sql
SELECT * FROM cypher('my_graph', $$ MATCH (n) RETURN n $$) AS (n agtype);
```

## Vector casts and operators inside Cypher

From the regression test, these casts are supported:

- `"[1.22,2.22,3.33]"::vector`
- `[1.22,2.22,3.33]::vector`
- `...::vector(3)` (dimension constrained)

And ordering by cosine distance works via the `<=>` operator:

```sql
SELECT * FROM cypher('graph', $$
  MATCH (m:Movie)
  RETURN m.title
  ORDER BY m.embedding::vector(4) <=> [-0.1, 0.2, 0.3, -0.4]::vector(4)
  ASC LIMIT 10
$$) AS (title agtype);
```

## Vector indexes (HNSW) on AGE vertex properties

The regression test demonstrates creating a HNSW index on a vertex property by building an expression that extracts the property and casts it to `vector(N)`:

1) Find the internal schema OID for the AGE graph
2) `CREATE INDEX ... USING hnsw ((... )::vector(N)) vector_cosine_ops;`

This is the core recipe to make vector similarity in AGE fast.

**Comment by nicholasveloso** [comment]
@webcoderz am I right in interpreting that this only work woth 3-dimension vectors? Or does it it support N-mrnsional vectors?

In ML, AI and Gen AI, vectors sizes are usualy in the order of tihousands dimension. For example,thr LLMs embedding vectors are in the order of 1500+ dimensions for smaller model, or 3000+ dimensions for larger models.

If this method works with N-domensional vectors, it would be a great addition to the Apache AGE project and its community. And maybe it finally opens up the portunity to use AGE as graph database for both of my startups. 

We are looking for an OSS graph database that can de as competitive as Neo4j a GraphDB and AWS Neptune, but thus far ageT was the closest databe to what we need, but the lack of ways for using vectors and vector embeddings is the only thing that is a d ql breaker for us. We already use pgvector as our vectorstore so integrating it with AGE would be the best case a scenario. Specialy if we eventually manage to generate graph embeddings using AGE ans pgvector.

King regards,

**Comment by webcoderz** [comment]
> [@webcoderz](https://github.com/webcoderz) am I right in interpreting that this only work woth 3-dimension vectors? Or does it it support N-mrnsional vectors?
> 
> In ML, AI and Gen AI, vectors sizes are usualy in the order of tihousands dimension. For example,thr LLMs embedding vectors are in the order of 1500+ dimensions for smaller model, or 3000+ dimensions for larger models.
> 
> If this method works with N-domensional vectors, it would be a great addition to the Apache AGE project and its community. And maybe it finally opens up the portunity to use AGE as graph database for both of my startups.
> 
> We are looking for an OSS graph database that can de as competitive as Neo4j a GraphDB and AWS Neptune, but thus far ageT was the closest databe to what we need, but the lack of ways for using vectors and vector embeddings is the only thing that is a d ql breaker for us. We already use pgvector as our vectorstore so integrating it with AGE would be the best case a scenario. Specialy if we eventually manage to generate graph embeddings using AGE ans pgvector.
> 
> King regards,

this is made from the regression tests that already live in their repo from a recent PR they merged in, im using it with like 384 n_dim embeddings right now, this is just a WIP on some of the notes ive written so far about it, im also in the progress of making an ORM for it all as well, and when i write some tests and validate it more ill link it .

**Comment by jrgemignani** [comment]
@nicholasveloso I'm not sure I understand. AGE works with pgvector. Could you provide an example of your question or issue?

**Comment by webcoderz** [comment]
```bash
pip install age_search
```
has BM25 support via pg_search as well as pgvector and age :). enjoy!
